Remove type provider security dialog and use custom icon for type provider assembly reference

enricosada opened the pull request:
ref #378 

- [ ] new icon ( @KevinRansom )
- [X] update icon
- [X] remove security dialog
- [x] remove tools options page
- [x] cleanup unused code

I was hasty about icon design, sry.

This is a pr to enable a custom icon based of assembly type (normal or type provider), but the custom icon will come from the VS branding team.

by @KevinRansom (see below)

> Just to set expectations.
Whilst I think it's interesting to scout ideas, and to produce prototypes, Icons are part of branding and VS branding is not owned by the feature team (us), so the outcome of this effort will likely serve as input to what the VS branding team does.

amazingant commented:
Cross-posting from #378, my 2¢ is to somehow base the icon off of the F# logos. My initial preference would be to just globally use the [blue logo][1], but it might look better to use the blue logo with Visual Studio's light and blue themes, but to use one of the [dark][2] [logos][3] with the dark theme.

All three for comparison (the empty space below is the light logo, and is basically white; preview it over something dark):

![blue logo][1]

![light logo][2]

![dark logo][3]

[1]: http://fsharp.org/img/logo/fsharp128.png
[2]: http://fsharp.org/img/logo/fsharp100mclight.png
[3]: http://fsharp.org/img/logo/fsharp100mcdark.png

forki commented:
@ReedCopsey as an official fsharp.org spokesperson, can you please comment on possible license issues?

enricosada commented:
16x16 is really small, with fsharp logo ( resized from http://fsharp.org/favicon.ico ) is like 
![image](https://cloud.githubusercontent.com/assets/147243/7613416/e5c2c9a4-f990-11e4-88e7-880120e99a0a.png)

enricosada commented:
added the placeholder, download e53ce24 to design and test the icon inside visualstudio

the placeholder is last spot in imagelist, and the icon is applied to all assemblies, so any project is ok

KevinRansom commented:
Just to set expectations.

Whilst I think it's interesting to scout ideas, and to produce prototypes, Icons are part of branding and VS branding is not owned by the feature team (us),  so the outcome of this effort will likely serve as input to what the VS branding team does.

Thanks

Kevin

latkin commented:
+1 to @KevinRansom, we can't just come up with an icon of our choosing. Lots of legal, globalization, and marketing factors to consider. It needs to come from our dedicated art/branding people, or at least re-use an already-approved icon used elsewhere in VS.

forki commented:
Since we all want to get it in asap we should consider to just change the color of the existing icon in order to formally fulfill the requirement of different icon. The beautification process can be done separately.

enricosada commented:
waah, these commit are for master branch not fsharp4 :cry: i'll fix this

Now the assembly reference icon is changed ( using old logic ) only after a build. 
I'll enable the other old path for dialog, and use that too for change icon (so icon change after add reference)

I removed pretty much all code/resources, splitted in multiple commit for easier diff

i tried to remove the type providers option page ( Options -> Tools -> Visual F# -> Type Providers ), removing

- the type TypeProviderToolsOptionsPage
- the ProvideOptionPage vsintegration/src/vs/FsPkgs/FSharp.Project/FS/Project.fs

but no luck, the menu is always there, only contains an error message about page creation.
So now there is the menu, but the page is empty. Can you remove it or we need to add a message?

enricosada commented:
ok, done. Rebased to fsharp4 and removed the dead code.

This pr use the old logic, but instead of show dialog the assembly ref icon is updated

I'll send another pr with more cleanup ( TcConfigBuilder.validateTypeProviders = ValidateTypeProviders msbuild prop etc ) to easier diff

as i said in my previous comment, how can i remove the type providers option page? I tried but no luck /cc @latkin @KevinRansom

enricosada commented:
@latkin @KevinRansom it's ok like this? you want less code removed for vs2015? or is too late?
I can leave the code, unused, to have a smaller commit. 
The commits are mostly deleted lines, low risk of regression

latkin commented:
This is almost surely going to be pushed to vNext (first 2015 update).  There is just a lot that needs to be pulled out (plus even more for complete elimination), we don't have an icon alternative on hand, and the options page is still an unknown.

This is great work, I just want to be transparent about what to expect.  We are at the point where all further changes need signoff from management, and something this big isn't likely to be accepted.

enricosada commented:
np @latkin i'll continue cleanup in this pr

ReedCopsey commented:
@forki Licensing would need to be discussed with the Board (though I suspect usage could be granted without problem), but I doubt the VS branding team would want to use this for this purpose...

enricosada commented:
removed all unused code, updated comment and cleanup.

### tools option page

i removed the tools options page code but the page continue to exists because the VF# extension ( not the EnableOpenSource extension ) add the registry entry `HKEY_CURRENT_USER\Software\Microsoft\VisualStudio\14.0FSharp_Config\ToolsOptionsPages\F# Tools\Type Providers` . 
I think a new build of vf# extension fix this problem /cc @KevinRansom

###  update assembly icon strategy

now the icon is updated after a build, but not if the project is already up-to-date (for example after  reopening of a solution). a new build update the icon, so np.
This remove the global displyDialog, recenteFile global and clean the code alot.

maybe we can remove the old flow or improve it. The flow is:

1. use fsc code to compile
2. use a function to publish type provider assemblies path
3. exit after assemblies are imported, before typecheck

but i need an hint /cc @dsyme @latkin @KevinRansom

latkin commented:
I'm working on getting this merged. It will be nice to say goodbye to the TP security dialog  :wave:

latkin commented:
Worked through this today and came to the following:
- Changes in this PR do a fine job of removing the TP dialog, that part is ok
- It's simple to remove the options menu page for TP approvals with a small setup change (remove a reg key)
- It's not entirely easy to retrofit the old "pop a dialog" codepaths to now mark a reference node as containing a type provider
  - TPs *can* be reliably checked at the point of launching a full build.  That part is implemented in this PR.
  - But what I would expect as a user, and what's not in this PR, is for the reference node to be marked properly at design-time, right when I add the reference.  Making this work reliably is a bit hard*.  Without this bit, you have to wait until you do a full build before references are updated.

So I propose we scope this PR down to simply "removing the dialog," ignoring the "mark a reference node as containing TPs" feature for now.  That's low-risk and achieves the primary goal of ditching the dialog.

I'll see what I can do to make the marking feature more robust, and open a separate PR for that.

`*` When the language service compiler encounters a type provider, the current mechanism for tying that back to a particular VS project is `GlobalsTheLanguageServiceCanPoke.theMostRecentFileNameWeChecked`. It's assumed that whatever project contains the last-processed file must also contain the just-encountered TP reference. From some quick testing, this is actually not a safe assumption - I found that the last-processed file was sometimes in a totally different project.  When we were just popping a dialog, it didn't really matter if this was inaccurate, but now it does.

latkin commented:
If [`ProcessCommandLineArgsAndImportAssemblies`](https://github.com/Microsoft/visualfsharp/blob/6a68e46d73763220cc9949e9540cbca0ecf9776a/src/fsharp/fsc.fs#L456) is cheap, then we could make the feature work reliably by simply removing the `if actuallyBuild then` condition [here](https://github.com/Microsoft/visualfsharp/blob/6a68e46d73763220cc9949e9540cbca0ecf9776a/vsintegration/src/vs/FsPkgs/FSharp.Project/FS/Project.fs#L1675-L1702).

Right now, adding a reference triggers the `Compile` method, but `actuallyBuild` is false so we don't get full references computation.

enricosada commented:
thx @latkin! i think the old code should be axed completly, is a global hack, and need lot of work for know in what projects the assembly is.
Is better to have a new clean method on the compiler service, to ask if an assembly is a type provider and maybe return the name of the type providers contained in the assembly (to add child nodes on the reference to easier discover type providers?). A new issue anyway

dsyme commented on src/fsharp/fscopts.fs:
```diff
@@ -444,7 +444,6 @@ let testFlag tcConfigB =
 // not shown in fsc.exe help, no warning on use, motiviation is for use from VS
 let vsSpecificFlags (tcConfigB: TcConfigBuilder) = 
   [ CompilerOption("vserrors", tagNone, OptionUnit (fun () -> tcConfigB.errorStyle <- ErrorStyle.VSErrors), None, None);
-    CompilerOption("validate-type-providers", tagNone, OptionUnit (fun () -> tcConfigB.validateTypeProviders <- true), None, None);
```
I do wonder if this flag should be left in and ignored.  

In theory the next version of the Visual F# fsc.exe compiler will only be used with the corresponding Visual F# Tools and FSharp.Build.dll.  But perhaps somehow the compiler will be invoked via the old tools, or an open source version of the tools, or some other tool that has copied out the command-line settings from the Visual F# Tools invocations of fsc.exe (which includes this flag).  Perhaps just ignoring the option would be better.

enricosada replied:
set as deprecated and noop? Because is a minor version (oob), no breaking changes
so if someone use it, get the warning but no error (build failed)
Next version (master?), cleanup and remove all deprecated commands?

latkin replied:
@dsyme yes I thought about that a bit, but dismissed it as it would need to be some kind of unsupported IDE setup... but I agree it's cheap and simple to keep it with a no-op so that we never have to worry.  Anyone is welcome to send that in a PR.
